clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2021 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import * as vscode from "vscode";
+import { Selection } from "vscode";
+import { PUNCH_CARD } from "../constants";
+import { ProfileService } from "../services/ProfileService";
+
+const ORIGINAL_TAB_COMMAND: string = "tab";
+const SMART_TAB_COMMAND: string = "cobol-lsp.smart-tab";
+
+const ORIGINAL_CMD_RIGHT_COMMAND: string = "cursorEnd";
+const SMART_NEXT_COLUMN_COMMAND: string = "cobol-lsp.next-column";
+
+const ORIGINAL_CMD_LEFT_COMMAND: string = "cursorHome";
+const SMART_PREV_COLUMN_COMMAND: string = "cobol-lsp.prev-column";
+
+const ORIGINAL_OUTDENT_COMMAND: string = "outdent";
+const SMART_OUTDENT_COMMAND: string = "cobol-lsp.smart-outdent";
+
+/**
+ * Class provides bind-to-command binding functionality
+ */
+abstract class SmartCommandProvider {
+    private context: vscode.ExtensionContext;
+    private originalCommandName: string;
+    private cobolCommandName: string;
+    private commandHolder: vscode.Disposable;
+
+    /**
+     * Constructor
+     * @param context is a extension context
+     * @param originalCommandName is a name for original command
+     * @param cobolCommandName is a name for cobol extension command
+     */
+    public constructor(context: vscode.ExtensionContext, originalCommandName: string, cobolCommandName: string) {
```

Comment:
You can use [parameter properties](https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties) here. So instead of:
```
abstract class SmartCommandProvider {
￼    private context: vscode.ExtensionContext;
    public constructor(context: vscode.ExtensionContext, ...) {
        this.context = context;
        ...
    }
```
it can be:
```
abstract class SmartCommandProvider {
    public constructor(private context: vscode.ExtensionContext, ...) {
    }
```